File: core-golang/api/event_handlers.go
package api

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"
	"vigil-core/logger"
	"vigil-core/notifier"
	"vigil-core/store"
)

// handlePostEvent receives a ChangeEvent from the scraper, saves it, and triggers notifications.
// POST /event
func (s *Server) handlePostEvent(w http.ResponseWriter, r *http.Request) {
	log := logger.Get()

	var event notifier.ChangeEvent
	if err := json.NewDecoder(r.Body).Decode(&event); err != nil {
		writeError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}

	if event.SiteID == "" || event.URL == "" {
		writeError(w, http.StatusBadRequest, "site_id and url are required")
		return
	}

	if event.DetectedAt == "" {
		event.DetectedAt = time.Now().UTC().Format(time.RFC3339)
	}

	log.Event(fmt.Sprintf("Change detected on %s! Dispatching notifications...", event.SiteID))

	// Save event to database
	detectedAt, _ := time.Parse(time.RFC3339, event.DetectedAt)
	dbEvent := &store.Event{
		SiteID:         event.SiteID,
		DetectedAt:     detectedAt,
		ScreenshotPath: event.ScreenshotPath,
	}
	if err := s.db.SaveEvent(dbEvent); err != nil {
		log.Error(fmt.Sprintf("Error saving event: %v", err))
		writeError(w, http.StatusInternalServerError, "failed to save event")
		return
	}

	// Trigger notifications in parallel
	go notifier.NotifyAll(s.tgCfg, s.emailCfg, event)

	writeJSON(w, http.StatusCreated, map[string]any{
		"status":   "ok",
		"event_id": dbEvent.ID,
		"message":  "event received, notifications dispatched",
	})
}

// handleTestNotification triggers a test notification to all configured channels.
// POST /test-notification
func (s *Server) handleTestNotification(w http.ResponseWriter, r *http.Request) {
	log := logger.Get()
	var timestamp string = time.Now().UTC().Format(time.RFC3339)

	event := notifier.ChangeEvent{
		SiteID:         "test-site",
		URL:            "https://example.com",
		DetectedAt:     timestamp,
		DiffSummary:    fmt.Sprintf("Test message generated from Vigil Dashboard. %s", timestamp),
		ScreenshotPath: "", // Optional
	}

	log.Info("Triggering test notification...")

	// Trigger notifications in parallel
	go notifier.NotifyAll(s.tgCfg, s.emailCfg, event)

	writeJSON(w, http.StatusOK, map[string]any{
		"status":  "ok",
		"message": "test notification dispatched",
	})
}

File: core/api/event_handlers.ts
import Elysia, { t } from "elysia";
import { DB } from "../store/db";
import { insertEvent } from "../store/events";
import { Notifier, ChangeEvent } from "../notifier/notifier";

export function eventHandlers(db: DB, notifier: Notifier) {
    return new Elysia()
        .post("/event", async ({ body }) => {
            try {
                const event = body as ChangeEvent;
                // 1. Save to DB
                insertEvent(db, {
                    site_id: event.site_id,
                    detected_at: event.detected_at,
                    screenshot_path: event.screenshot_path
                });

                // 2. Notify
                await notifier.notifyAll(event);

                return { status: "ok" };
            } catch (err: any) {
                return new Response(JSON.stringify({ error: err.message }), { status: 500, headers: { "Content-Type": "application/json" } });
            }
        }, {
            body: t.Object({
                site_id: t.String(),
                url: t.String(),
                screenshot_path: t.Optional(t.String()),
                diff_summary: t.String(),
                detected_at: t.String()
            })
        })
        .post("/test-notification", async () => {
            try {
                const testEvent: ChangeEvent = {
                    site_id: "test-site",
                    url: "https://example.com",
                    diff_summary: "This is a test notification from Vigil Core (Bun).",
                    detected_at: new Date().toISOString()
                };
                await notifier.notifyAll(testEvent);
                return { status: "sent test" };
            } catch (err: any) {
                return new Response(JSON.stringify({ error: err.message }), { status: 500, headers: { "Content-Type": "application/json" } });
            }
        });
}
